specs/WebResourceRequested-CustomScheme.md

```diff
@@ -0,0 +1,268 @@
+}
+
+[uuid(ac52d13f-0d38-475a-9dca-876580d6793e), object, pointer_default(unique)]
+interface ICoreWebView2StagingEnvironmentOptions : IUnknown {
+  /// Collection of custom scheme registrations
+  [propget] HRESULT CustomSchemeRegistrar(
+      [out, retval] ICoreWebView2StagingCustomSchemeRegistrationCollection** schemeRegistrar);
+}
+```
+
+## .NET API
+```c#
+namespace Microsoft.Web.WebView2.Core
+{
+    // Represents the registration of a custom scheme with the CoreWebView2Environment.
+    // This allows the WebView2 app to be able to handle
+    // WebResourceRequested event for requests with the specified scheme and
+    // be able to navigate the WebView2 to the custom scheme. Once the environment
+    // is created, the registrations are valid and immutable throughout the
+    // lifetime of the associated WebView2s' browser process and any WebView2
+    // environments sharing the browser process must be created with identical
+    // custom scheme registrations, otherwise the environment creation will fail.
+    // Any further attempts to register the same scheme will fail.
+    // The URIs of registered custom schemes will be treated similar to http URIs for their origins.
+    // They will have tuple origins for URIs with host and opaque origins for
+    // URIs without host as specified in https://html.spec.whatwg.org/multipage/origin.html
+    // Example:
+    // custom-scheme-with-host://hostname/path/to/resource has origin of custom-scheme-with-host://hostname
+    // custom-scheme-without-host:path/to/resource has origin of custom-scheme-without-host:path/to/resource
+    // For WebResourceRequested event, the request URIs and filter URIs with custom schemes
+    // will be normalized according to RFC3986.
+    runtimeclass CoreWebView2StagingCustomSchemeRegistration
+    {
+        // The name of the custom scheme to register.
+        String SchemeName { get; set; };
+
+        // Whether the scheme will treated as a secure context like https.
+        bool IsSecure { get; set; };
+
+        // If `isCorsEnabled` is set, the scheme will be allowed for requests subject to
+        // CORS rules as specified in https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS.
+        // If it is not set, CORS requests with this scheme will be blocked.
+        bool IsCorsEnabled([out, retval] { get; set; };
+
+        // If `isCspEnabled` is set, the scheme is allowed to be used for requests
+        // that are subject to CSP including from itself as specified in
+        // https://developer.mozilla.org/en-US/docs/Web/HTTP/CSP
+        // If not set, such CSP subject requests will be blocked.
+        bool IsCspEnabled { get; set; };
+
+        // List of origins that are allowed to use the scheme,
+        // subject to cross origin restrictions.
+        // Origins are specified as a string in the format of scheme://host:port.
+        // `*` wildcard can also be used on origins.
```

Comment:
Is the wildcard applied just like for CSP rules? Can we define it in terms of that if so?

If not, we should also define if wildcard can be used in the scheme and if it can show up multiple times.

Reply:
My plan is to make it a simple string match.